### ISSUE TITLE
Incorrect result when concatenating varchar and char

### ISSUE THREAD

**Issue by findepi** [trinodb/trino#10118]
Concatenating some value of `varchar(n)` with `char(m)` makes the varchar value space-padded.
The padding shouldn't be there.

```
trino> SELECT v3 || c3, format('[%s]', v3 || c3)
    -> FROM (VALUES (CAST('a' AS varchar(3)), CAST('c' AS char(3)))) t(v3, c3);
 _col0  |  _col1
--------+----------
 a  c   | [a  c  ]
(1 row)
```

cc @ebyhr @kasiafi @martint @sopel39

**Comment by findepi** [comment]
This might be related to https://github.com/trinodb/trino/commit/3d9d004bde77c847108ece06c727536669b6a640

**Comment by sopel39** [comment]
result would be of `char` type, right?, so it seems correct with current semantics

**Comment by martint** [comment]
> This might be related to 3d9d004

Yes, currently, the coercions between char and varchar are reversed, and the type for string literals is incorrect (i.e., it should be `char(n)`). I looked into what it would take to fix that last year, but didn't make much progress:

https://github.com/martint/trino/commit/4d4ec83f4f0a8a54e0a32fae5a55f196b14cc9d7

**Comment by findepi** [comment]
> the type for string literals is incorrect (i.e., it should be char(n)).

agreed.

> the coercions between char and varchar are reversed

yes, and fixing the coercion direction requires changing string literals' type

Still, we should be able to fixed concatenation under current char/varchar coercion rules.

@martint please confirm or disconfirm my assessment of what's expected here.

**Comment by martint** [comment]
Yes. Your understanding is correct. It should behave as if the `char` were being converted to a `varchar` and then concatenated with the other `varchar`. Additionally, the result of the concatenation should be `varchar(6)`.

**Comment by findepi** [comment]
Relates to #9031

**Comment by ericxiao251** [comment]
> Concatenating some value of `varchar(n)` with `char(m)` makes the varchar value space-padded. The padding shouldn't be there.
> 
> ```
> trino> SELECT v3 || c3, format('[%s]', v3 || c3)
>     -> FROM (VALUES (CAST('a' AS varchar(3)), CAST('c' AS char(3)))) t(v3, c3);
>  _col0  |  _col1
> --------+----------
>  a  c   | [a  c  ]
> (1 row)
> ```
> 
> cc @ebyhr @kasiafi @martint @sopel39

Hi folks, I am hoping to tackle this issue :). 

What should the correct result be, `ac` with a type `char(6)`?

**Comment by ericxiao251** [comment]
### Findings

Hi folks,

@damnMeddlingKid have done some investigations into the current behavior of concatenating a `char` and a `varchar` in Trino and other systems (Postgres).

Current Behaviour in Trino (version 384): 

| col a          | col b          | result  | result datatype |
|----------------|----------------|---------|-----------------|
| 'a' char(3)    | 'c' char(2)    | [a  c ] | char(5)         |
| 'a' char(3)    | 'c' varchar(2) | [a  c ] | char(5)         |
| 'a' varchar(3) | 'c' char(2)    | [a  c ] | char(5)         |
| 'a' varchar(3) | 'c' varchar(2) | [ac]    | varchar         |

Behaviour in PostgreSQL (version 13.3):
| col a          | col b          | result  | result datatype |
|----------------|----------------|---------|-----------------|
| 'a' char(3)    | 'c' char(2)    | [ac]    | text            |
| 'a' char(3)    | 'c' varchar(2) | [ac]    | text            |
| 'a' varchar(3) | 'c' char(2)    | [ac]    | text            |
| 'a' varchar(3) | 'c' varchar(2) | [ac]    | text            |

-----
### Initial Questions:

1. From the above results, are there some Trion configurations we need to account for? I saw some backward compatibility configurations around type coercion but cannot find those again.
2. What's the expected result and result datatype?
3. Where would the type coercion code chunks live?

-----

### Proposed solution
| col a          | col b          | result  | result datatype |
|----------------|----------------|---------|-----------------|
| 'a' char(3)    | 'c' char(2)    | [ac] | char(5)         |
| 'a' char(3)    | 'c' varchar(2) | [ac] | char(5)         |
| 'a' varchar(3) | 'c' char(2)    | [ac] | varchar         |
| 'a' varchar(3) | 'c' varchar(2) | [ac]    | varchar         |

Our proposed solution would be:
1. Remove the spaces between and after the characters in the `result`.
2. The resulting data type will be the data type of the first column.

We also noticed that when concatenating two varchars, the maximum length precision of the varchar was lost - if this is a bug, we can can fix it in a follow up PR (filed in this ticket: https://github.com/trinodb/trino/issues/12827).

**Comment by ericxiao251** [comment]
Hey @findepi / @martint, do you have any feedback on the above [comment](https://github.com/trinodb/trino/issues/10118#issuecomment-1154081828) I made? Would love to make a contribution to Trino but want to understand what the correct path forward is :).

**Comment by findepi** [comment]
> 
'a' varchar(3) | 'c' char(2) | [ac] | varchar
-- | -- | -- | --

That's not correct. 
There is no such thing as `'c' char(2)` (using your notation)
It's a short-hand for `'c⎵' char(2)` (`⎵` stands for a space)
So concatenation result should be `ab⎵`.

As to the result _type_, `varchar` is correct, but `varchar(5)` would be preferred.

> 
'a' char(3) | 'c' varchar(2) | [ac] | char(5)
-- | -- | -- | --

Similar to the above, the lhs value actually is `a⎵⎵` and ths value is `c`.
It must not be of type `char(n)` -- we must not _add_ spaces during the concatenation.
The result should be `a⎵⎵c`. It needs to be of type `varchar(5)` (or `varchar`).
